src/qw/mergedoc.py

```diff
@@ -0,0 +1,301 @@
+""" Merges data into output documents """
+from collections.abc import Iterable
+import copy
+import docx
+from loguru import logger
+from typing import Self, List, Dict, Any
+
+from md import markdown_to_plain_text
+from docsection import DocSection, DocSectionParagraphReplacer
+
+NONE_PARAGRAPH = "*None.*"
+
+class MergeData:
+    """
+    Data to be merged into the document.
+
+    The basic format is a dictionary of object names to either
+    a dictionary of strings to values, or a list of dictionaries
+    of strings to values.
+
+    So you might have:
+    {"ObjectName": {
+        "id": 45,
+        "name": "MyObject",
+        "description": "The object that is mine"
+    }}
+    which would provide three values called "ObjectName.id",
+    "ObjectName.name" and "ObjectName.description".
+
+    On the other hand, you could also have:
+    {"ObjectName": [{
+        "id": 45,
+        "name": "MyObject",
+        "description": "The object that is mine"
+    }, {
+        "id": 46,
+        "name": "YourObject",
+        "description": "The object that is yours"
+    }, {
+        "id": 47,
+        "name": "TheirObject",
+        "description": "The object that is theirs"
+    }]}
+    This also provides the same three values, but there are now
+    three of each.
+
+    If you put one or more of these in a paragraph in your template
+    document, this paragraph (plus, if it is a heading then all the
+    paragraphs beneath it including subheadings) will be copied
+    enough times to hold each value.
+
+    Some of these objects might refer to each other. So, if an
+    object "MyObject" has a value "OtherObjectName": "123" and there
+    is an object called "OtherObjectName", the value is assumed to
+    refer to the "OtherObjectName" objects whose "id" key is "123".
+
+    We can use this for nested loops. In this case if you have a
+    section of the document with fields from "OtherObjectName" with
+    "id": "123" then the fields referring to "MyObject" will only
+    select those "MyObject" objects whose "OtherObjectName" values
+    match "123".
+
+    In this way we can have a heading advertising "OtherObjectName"
+    and then subheadings advertising "MyObject" and each MyObject
+    subheading will be placed nicely under the correct
+    "OtherObjectName" heading.
+
+    This will also work with nested lists. It will eventually work
+    with tables, too, but that might be harder to describe and
+    harder to implement.
```

Comment:
Should we just use excel for tables? May fit more naturally in how users would want to interact and edit them?

Reply:
Yeah, could do. I'd really like to have tables but that decision has to be for later.